### PR TITLE
(feat) 030: React - define homework

### PR DESCRIPTION
**Pull request by unfalse**
This is just a rip-off from the RSSchool React course. I removed react-router and other things.

### CODE REVIEW

**Comment by unfalse** [comment]
TODO:
Remove git, CRA, eslint, prettier, tests, typescript
Add codesandbox, marks, merge homework from rsschool together module 1-6, evaluation criteria

File: react/homework.md [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+## React. Components
+
+### What should be done:
+
+1. Create a new app using the React template on https://codesandbox.io
+2. Use class components. Using hooks is forbidden at this stage.
+3. Draw the Search Bar and Cards in the div tag with id "search-bar". Try to make it simple.
+* examples of Search Bar:\
+https://studio.uxpincdn.com/studio/wp-content/uploads/2020/09/BlogHeader_SearchBar_1200x600.png \
```
Comment (Dzmitry-Klokau) [inline react/homework.md:13]: could you place images in out repo?

File: react/homework.md [hunk 2]
```diff
@@ -0,0 +1,127 @@
+# React homework
```
Comment (Dzmitry-Klokau) [inline react/homework.md:1]: Could you combine the sub-homework into one big homework.

**Comment by Dzmitry-Klokau** [comment]
please link your branch to the GitHub issue

**Comment by unfalse** [comment]
> please link your branch to the GitHub issue

done

File: react/homework.md (hunk 1 above)
Comment (unfalse) [reply react/homework.md]: done

File: react/homework.md (hunk 2 above)
Comment (unfalse) [reply react/homework.md:1]: done

File: react/homework.md [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+  **The usage of Redux or any other solutions for state management is prohibited.**
+
+  **The usage of any libraries with components or form libraries is prohibited.** 
+
+## Steps
+
+1. Create a new app using the React template on https://codesandbox.io
+2. Use function components with hooks.
+3. Draw the Search Bar and Cards in the div tag with id "search-bar".
+4. Input value should be saved to LocalStorage during component’s unmount. During the initialization pick the value from LocalStorage and show it.
```
Comment (Dzmitry-Klokau) [inline react/homework.md:21]: please add a tick how to work with LocalStorage
probably some YouTube link

File: react/homework.md [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+     ```
+     text input - name, surname, zip-code;
+     date input - birthday, date of delivery;
+     dropdown/select - list of countries, list of states (User can choose only one element from the list)
+     checkbox - "I consent to my personal data" field, list of extra presents (User can choose several items from the list)
+     switcher - male/female, "I want to receive notifications about promo, sales, etc." / "I don’t want to receive notifications about promo, sales, etc."
+     file upload - profile picture
+     ```
+
+   After the Submit button is clicked, all data from the form has to be displayed as Card in the cards list below the form. The confirmation message (with information that the data has been saved) should be displayed, after that form has to be cleared.
```
Comment (Dzmitry-Klokau) [inline react/homework.md:57]: please add a tip how to do it
use modal? alert?
link to docs

File: react/homework.md [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+* examples of Cards:
+![example of Cards](images/behance.jpeg)
+![example of Cards](images/rightmove.jpeg)
+![example of Cards](images/awwwards.jpeg)
+
+4. Create a component with form in a new div tag with id "form-component".
+5. Form should be implemented using uncontrolled components.
+6. Add functionality to your app:
+   +  Collect information through a form
+   +  You can collect any type of information, but form has to include at least one control of the following types:
```
Comment (Dzmitry-Klokau) [inline react/homework.md:39]: single html as an example?

File: react/homework.md [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+   
+   The quantity of cards should equal the number of form submissions (if the user filled in and submitted the form five times, then five cards have to be displayed under the form).
+
+   ### **Form validation**
+   
+   Validation should happen after the **Submit** button was clicked. In case some field was filled incorrectly, then error messages should be displayed under it. The card mustn’t be created until the user fixes all errors.
+   
+   After changing the value of erroneous input, the reset should happen only for this particular field, but not for the others.
+   
+   Submit button should be disabled at initialization (before the first typing), and until the user fixes all errors.
```
Comment (Dzmitry-Klokau) [inline react/homework.md:67]: can we create one more html file with examples of html elements and their designs in different states? buttons, inputs, download indicator

File: react/homework.md [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+All logical parts should be extracted into components.
+   
+All data must be stored in a local state of the component.
+   
+Components should not make calls to APIs.
+
+7. Choose an API.
+   There are several recommended APIs. But you may choose any other API that you prefer if it supports search, pagination and sorting.
```
Comment (Dzmitry-Klokau) [inline react/homework.md:76]: can be postponed 
but just and idea
can we create our own api?

File: react/homework.md [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+   - https://the-one-api.dev/documentation
+   - https://rickandmortyapi.com/documentation/
+   - https://developers.themoviedb.org/3/getting-started/introduction
+   - https://pipedream.com/apps/swapi
+   - https://github.com/public-apis/public-apis
+     **The API should support searching, pagination and sorting. Although you don’t need it at this stage, but there’s a good chance it might come in handy later.**
+8. Use the search page from the first part and rewrite it. Split it into 2 logical parts:
+   - **Search bar.** Enter text -> Press Enter -> Send the request to API with the searching parameters -> the list of results is updating
+   - **The result list of searching**: display the data that the API query returned using Cards.
+9. A customer decided to change requirement. Now every element of the list has to show a small piece of information. After the click on the element we have to show all the  available information about the element in the modal window. The modal window should be closed by clicking on the cross button in the upper right corner or by clicking on a page outside the modal window. When we open a modal window the page should be covered with overlay.
```
Comment (Dzmitry-Klokau) [inline react/homework.md:87]: link to modal docs

File: react/homework.md [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+All the logical parts must be in separated components.
+**User-friendly interface with the downloading indicator and with notification messages in a case if something goes wrong or some information cannot be found is warmly welcome.**
+
+11. Retain the functionality and **rewrite the form** with library **React Hook Form**. 
+12. Implement state management solution using useReducer and Context API for search results and form data saving.
+13. Add to global state
+    - **switches for sorting** (at least three sorting options). Select a parameter -> send a request to the API with the entered parameter -> the list of search results is updated
+    - **switches for pagination.** It should be possible to select the number of results per page and the number of the page we are currently on, the total number of pages. Select a parameter -> send a request to the API with the entered parameter -> the list of search results is updated
+14. The customer has changed the requirements for the output:
```
Comment (Dzmitry-Klokau) [inline react/homework.md:98]: if we have no navigation library...probably we can remove this point

File: react/homework.md [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+1. Search bar + components – **2 points**
+2. Form + Cards - **2 points**
+3. API and modal window - **2 points**
+4. Custom State Management - **2 points**
+5. Sorting and pagination - **2 points**
+6. Search details - **2 points**
+
+If one of the items mentioned above is not accomplished – grade for the task is **0 points**. If there are direct DOM manipulations – **0 points**.
+
+### Theory
```
Comment (Dzmitry-Klokau) [inline react/homework.md:117]: please filter this list
we want to have only ENG sources

**Review by Dzmitry-Klokau — CHANGES_REQUESTED** [review]

File: react/homework.md [hunk 11]
```diff
@@ -0,0 +1,135 @@
+# React homework
+
+## Overview
```
Comment (Dzmitry-Klokau) [inline react/homework.md:3]: Could you add ErrorBoundary as an example of class component usage?

File: react/homework.md (hunk 3 above)
Comment (unfalse) [reply react/homework.md:22]: done

File: react/homework.md (hunk 10 above)
Comment (unfalse) [reply react/homework.md:124]: done

File: react/homework.md (hunk 4 above)
Comment (unfalse) [reply react/homework.md]: done

**Pull request closed without merging**